Inconsistent expo-sqlite/next

### Minimal reproducible example
https://github.com/abduraufsherkulov/expo-sqlite
-

### What platform(s) does this occur on?

Android

### Did you reproduce this issue in a development build?

Yes

### Summary

Updated my codebase to use expo-sqlite/next, however it seems API is working inconsistently. Given I do make requests multiple times to SELECT, it is throwing:

```
[Error: Call to function 'NativeDatabase.prepareAsync' has been rejected.
Caused by: java.lang.NullPointerException: java.lang.NullPointerException]
```

Video (please bear with me, it stops working only after making requests multiple times):
https://github.com/expo/expo/assets/17489128/f9dafaaf-5f5e-4c60-a797-1af8e507c777


### Environment

```
  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.4
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.9.0 - ~/.nvm/versions/node/v20.9.0/bin/node
      Yarn: 1.22.11 - /usr/local/bin/yarn
      npm: 10.1.0 - ~/.nvm/versions/node/v20.9.0/bin/npm
      Watchman: 2024.01.22.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.14.3 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.2, iOS 17.2, macOS 14.2, tvOS 17.2, visionOS 1.0, watchOS 10.2
      Android SDK:
        API Levels: 23, 28, 29, 30, 31
        Build Tools: 28.0.3, 29.0.2, 31.0.0
        System Images: android-31 | Google APIs ARM 64 v8a
    IDEs:
      Android Studio: 2020.3 AI-203.7717.56.2031.7784292
      Xcode: 15.2/15C500b - /usr/bin/xcodebuild
    npmPackages:
      expo: ~50.0.11 => 50.0.11 
      expo-router: ~3.4.8 => 3.4.8 
      react: 18.2.0 => 18.2.0 
      react-native: 0.73.4 => 0.73.4 
    npmGlobalPackages:
      eas-cli: 7.4.0
      expo-cli: 6.3.10
    Expo Workflow: managed
```

### Expo Doctor Diagnostics

-

--- a/packages/expo-sqlite/CHANGELOG.md
+++ b/packages/expo-sqlite/CHANGELOG.md
@@ -9,6 +9,7 @@
 ### 🐛 Bug fixes
 
 - Enabled [FTS](https://www.sqlite.org/fts3.html) and [FTS5](https://www.sqlite.org/fts5.html) for SQLite. ([#27738](https://github.com/expo/expo/pull/27738) by [@kudo](https://github.com/kudo))
+- Fixed `NullPointerException` on Android when opening the same database multiple times. ([#27748](https://github.com/expo/expo/pull/27748) by [@kudo](https://github.com/kudo))
 
 ### 💡 Others
 

--- a/packages/expo-sqlite/android/src/main/java/expo/modules/sqlite/NativeDatabase.kt
+++ b/packages/expo-sqlite/android/src/main/java/expo/modules/sqlite/NativeDatabase.kt
@@ -3,9 +3,15 @@
 package expo.modules.sqlite
 
 import expo.modules.kotlin.sharedobjects.SharedRef
+import java.util.concurrent.atomic.AtomicInteger
 
 internal class NativeDatabase(val databaseName: String, val openOptions: OpenDatabaseOptions) : SharedRef<NativeDatabaseBinding>(NativeDatabaseBinding()) {
   var isClosed = false
+  private val refCount = AtomicInteger(1)
+
+  internal fun addRef() {
+    refCount.incrementAndGet()
+  }
 
   override fun equals(other: Any?): Boolean {
     return other is NativeDatabase && this.ref == other.ref
@@ -13,6 +19,9 @@ internal class NativeDatabase(val databaseName: String, val openOptions: OpenDat
 
   override fun deallocate() {
     super.deallocate()
-    this.ref.close()
+    val shouldClose = refCount.decrementAndGet() <= 0
+    if (shouldClose) {
+      this.ref.close()
+    }
   }
 }

--- a/packages/expo-sqlite/android/src/main/java/expo/modules/sqlite/SQLiteModuleNext.kt
+++ b/packages/expo-sqlite/android/src/main/java/expo/modules/sqlite/SQLiteModuleNext.kt
@@ -60,6 +60,7 @@ class SQLiteModuleNext : Module() {
 
           // Try to find opened database for fast refresh
           findCachedDatabase { it.databaseName == databaseName && it.openOptions == options && !options.useNewConnection }?.let {
+            it.addRef()
             return@Constructor it
           }
 

